test/e2e/dr/quorum-restore.go

```diff
@@ -0,0 +1,330 @@
+			node := &masterNodes.Items[i]
+			etcdName := fmt.Sprintf("etcd-%d.%s", i, domain)
+			masterIP := ""
+			for _, address := range node.Status.Addresses {
+				if address.Type == "InternalIP" {
+					masterIP = address.Address
+					break
+				}
+			}
+			if node.GetName() == survivingNodeName {
+				survivingMasterIP = masterIP
+			}
+			updateDNS(domain, etcdName, masterIP)
+		}
+
+		imagePullSecretPath := getPullSecret(oc)
+		defer os.Remove(imagePullSecretPath)
+		runPodSigner(oc, survivingNodeName, imagePullSecretPath, proxy)
+
+		e2e.Logf("Restore etcd on remaining masters")
+		setupEtcdEnvImage := getImagePullSpecFromRelease(oc, imagePullSecretPath, "setup-etcd-environment")
+		kubeClientAgent := getImagePullSpecFromRelease(oc, imagePullSecretPath, "kube-client-agent")
+		for i := range masterNodes.Items {
+			node := &masterNodes.Items[i]
+			masterDNS := ""
+			for _, address := range node.Status.Addresses {
+				if address.Type == "InternalDNS" {
+					masterDNS = address.Address
+					break
+				}
+			}
+			if masterDNS == survivingNodeName {
+				e2e.Logf("Skipping node as its the surviving master")
+				continue
+			}
+			runViaBastionSSH(masterDNS, proxy,
+				fmt.Sprintf("sudo -i env SETUP_ETCD_ENVIRONMENT=%s KUBE_CLIENT_AGENT=%s /bin/bash -x /usr/local/bin/etcd-member-recover.sh %s \"etcd-member-%s\"",
+					setupEtcdEnvImage, kubeClientAgent, survivingMasterIP, node.GetName()))
+		}
+
+		e2e.Logf("Wait for etcd pods to become available")
+		_, err = exutil.WaitForPods(
+			oc.AdminKubeClient().CoreV1().Pods("openshift-etcd"),
+			exutil.ParseLabelsOrDie("k8s-app=etcd"),
+			exutil.CheckPodIsReady,
+			expectedNumberOfMasters,
+			10*time.Minute,
+		)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		scaleEtcdQuorum(oc, expectedNumberOfMasters)
+
+		e2e.Logf("Remove etcd signer")
+		err = oc.AdminKubeClient().CoreV1().Pods("openshift-config").Delete("etcd-signer", &metav1.DeleteOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		restartSDNPods(oc)
+		waitForMastersToUpdate(oc, mcps)
+		waitForOperatorsToSettle(coc)
+	})
+})
+
+func scaleEtcdQuorum(oc *exutil.CLI, replicas int) {
+	e2e.Logf("Scale etcd-quorum-guard to %d replicas", replicas)
+	etcdQGScale, err := oc.AdminKubeClient().AppsV1().Deployments("openshift-machine-config-operator").GetScale("etcd-quorum-guard", metav1.GetOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+	etcdQGScale.Spec.Replicas = int32(replicas)
+	_, err = oc.AdminKubeClient().AppsV1().Deployments("openshift-machine-config-operator").UpdateScale("etcd-quorum-guard", etcdQGScale)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	etcdQGScale, err = oc.AdminKubeClient().AppsV1().Deployments("openshift-machine-config-operator").GetScale("etcd-quorum-guard", metav1.GetOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(etcdQGScale.Spec.Replicas).To(o.Equal(int32(replicas)))
+}
+
+func runPodSigner(oc *exutil.CLI, survivingNodeName, imagePullSecretPath, proxy string) {
```

Comment:
aren't customers always going to have to do this?  Why wouldn't it be part of your script?

Reply:
Correct, this is [part of the doc](https://docs.openshift.com/container-platform/4.1/disaster_recovery/scenario-1-infra-recovery.html), "Create the signer Pod using the file that was generated"